research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+
+    Returns:
+        int: The number of spatial dimensions
+    """
+    if nnunet_config == "2d":
+        return 2
+    elif nnunet_config in ["3d_fullres", "3d_lowres", "3d_cascade_fullres"]:
+        return 3
+    else:
+        raise TypeError(f"Got unexpected nnunet config: {nnunet_config}")
+
+
+def convert_ds_list_to_dict(
+    tensor_list: Union[List[torch.Tensor], Tuple[torch.Tensor]], num_spatial_dims: int
+) -> Dict[str, torch.Tensor]:
+    """
+    Converts a list of torch.Tensors to a dictionary. Names the keys for
+    each tensor based on the spatial resolution of the tensor and it's
+    index in the list. Useful for nnUNet models with deep supervision where
+    model outputs and targets loaded by the dataloader are lists. Assumes the
+    tensors have a batch dimension.
+
+    Args:
+        tensor_list (List[torch.Tensor]): A list of tensors, usually either
+            nnunet model outputs or targets, to be converted into a dictionary
+        num_spatial_dims (int): The number of spatial dimensions. Assumes the
+            spatial dimensions are last
+    Returns:
+        Dict[str, torch.Tensor]: A dictionary containing the the tensors as
+            values where the keys are 'i-XxYxZ' where i was the tensors index
+            in the list and X,Y,Z are the spatial dimensions of the tensor
+    """
+    # Convert list of targets into a dictionary
+    tensors = {}
+
+    for i, tensor in enumerate(tensor_list):
+        # generate a key based on the spatial dimension and index
+        key = str(i) + "-"
+        for i in range(num_spatial_dims, 0, -1):
+            key += str(tensor.shape[-i]) + "x"
+        key = key[:-1]  # remove the trailing 'x'
+
+        tensors.update({key: tensor})
+
+    return tensors
+
+
+def convert_ds_dict_to_list(tensor_dict: Dict[str, torch.Tensor]) -> List[torch.Tensor]:
+    """
+    Converts a dictionary of tensors back into a list so that it can be used
+    by nnunet deep supervision loss functions
+
+    Args:
+        tensor_dict (Dict[str, torch.Tensor]): Dictionary containing
+            torch.Tensors. The key values must start with 'X-' where X is an
+            integer representing the index at which the tensor should be placed
+            in the output list
+
+    Returns:
+        List[torch.Tensor]: A list of torch.Tensors
+    """
+    tensor_list = []
+    for i in range(len(tensor_dict.keys())):  # keys start with index
+        key = [k for k in tensor_dict.keys() if k.startswith(str(i))]
+        assert (
+            len(key) == 1
+        ), f"found more than one key that started with {i}. \
+            Was expecting the first character to be the index in the \
+            list from which the predictions were derived. You're \
+            using an nnunet model with deep supervision right?"
+
+        tensor_list.append(tensor_dict[key[0]])
+
+    return tensor_list
+
+
+def is_valid_config(val: str) -> bool:
+    return val in list(get_args(nnUNetConfig))
```

Comment:
I like this idea, however i just noticed that I can't start the names of the items in enum with numbers, so i would have to prefix everything with 'cfg' or something which makes it less readable. Not sure if that defeats the purpose or if it is still useful to structure it this way

Reply:
Right...I forgot about that. I think it's still useful, though less satisfying, since it adds structure to the config options and can be imported elsewhere etc. A couple other options for this would be to spell out the number `TWO_D_..` and `THREE_D...` or `_2D...` and `_3D...`. None of them are perfect, but I think there is still benefit in the enum.